### PR TITLE
Display messages if JS or WebAssembly are not enabled.

### LINKED ISSUE #23: Display messages if JS or WASM are not enabled
The `index.html` file (yes there is absolutely no reason to use .xhtml^^) should contain sth like:

```html
<noscript>This website requires JavaScript.</noscript>
```

at the top of the website.

And in the JavaScript code we want to check if `typeof WebAssembly === 'object'` and otherwise do something like `alert("This website requires WebAssembly.")`.

### PR DESCRIPTION
**Pull request by Alessandro-Battiato**
Resolves #23  
Added HTML code to check if user's browser supports JavaScript and also added JS logic to alert the user if the browser does not
support WebAssembly

### CODE REVIEW

**Comment by not-my-profile** [comment]
Since you said that this is your first contribution to solve an issue, here some feedback:

* You can put `Resolves #<issue-number>.` in the PR description (where `<issue-number>` is the number of the GitHub issue you are solving) to make GitHub automatically link the PR to the issue (and merging the PR will automatically close the issue).
* Your commit implemented new features and also did some minor reformatting that was unrelated to the issue. Normally you want to separate formatting changes and function changes into separate commits so that they can be easily reviewed independently of each other (and also have descriptive commit messages attached). Here this doesn't really matter since the diff is so small but you may still want to try to separate these changes as a learning exercise.

  Regarding your formatting changes I don't really think the formatting changes in the HTML file improve anything. Regarding the JS file ... ideally this project would have some autoformatting set up that was checked in the CI ... currently we don't have that here.

* The commit message could be a bit more descriptive ... we don't just check if these are used but also alert the user ... so maybe `Display messages if JS or WebAssembly are not enabled`.

**Comment by Alessandro-Battiato** [comment]
First of all, I'm grateful you took your time to give me some feedback, so thanks for that :)

As for the issues you're pointing to, thanks again for the suggestion to include the issue number in the PR description, I didn't know that github worked like this!

Now I actually have a couple of questions to understand if there was a huge misunderstanding from my side regarding the issue but I need to give you context of what I've done.

Basically I added the no script in the HTML file as it appears only when your browser does not use JavaScript, and I thought this was the first task because indeed when disabling javascript in my browser's settings the message pops up. So question number one, if this is not the solution then what's the actual task?

Then regarding the formatting changes, could you explain in detail what you mean? If I understand correctly I should have first created the function logic and then commit it, and only then call it by also changing the original code?

Last question, I thought that the task about warning the user that the website uses the WebAssembly API should consist only in warning the user with an alert, and again what I've tested on my browser (specifically done this on Firefox though) was disabling the javascript options .wasm files and then the alert pops up immediately when the website is opened.

If this was not the solution and you mentioned an autoformatting setup using the CI, could you give me an example of what you meant?

Sorry if I'm asking too much but as you see I'm trying my best to learn and I'd be grateful if you again provided me more constructive criticism, which I'm grateful for :)

Have a nice day!

**Comment by not-my-profile** [comment]
>  thanks again for the suggestion to include the issue number in the PR description

You can edit the PR description by clicking on the three horizontal dots.

> I need to give you context of what I've done

You don't need to ... I can see exactly what you have done by viewing the diff of your PR (e.g. by clicking on the "Files changed" tab on this PR). And yes you managed to solve the issue :) You did what was described :)

> Then regarding the formatting changes, could you explain in detail what you mean?

Look at the diff of your PR or your commit. You will see changes like the following:
image
image

These are what I mean by formatting changes. It can very well be that you did not do them intentionally and your editor automatically formatted the file.

> you mentioned an autoformatting setup using the CI

Sorry I didn't want to confuse you. Setting up such a setup is out of the scope of the issue that you are solving. By autoformatting i meant it would be nice if we set up something like [Prettier](https://prettier.io/). GitHub has a [CI feature](https://docs.github.com/en/actions/automating-builds-and-tests/about-continuous-integration) that would let us check that the files are correctly autoformatted ... but this is as I said out of the scope of the issue ... don't worry about it :)

I hope that cleared things up ... have a nice day as well :)

Note that reworking a PR so that the commits contain precisely the changes you want requires some know how of the git command-line interface which can be a bit overwhelming at first but I do recommend learning it in the long run. E.g. you could run `git reset HEAD~` to undo the last commit while keeping your changes and then selectively stage the changes you want to commit (e.g. everything except the formatting changes) with `git add -p` ... and then commit with `git commit`. Git is very powerful if you start having PRs with multiple commits it's also worth it to learn `git rebase -i`, but for such a simple PR you don't need it. Once you have changed your commits as you want them you can force push them to the GitHub PR with `git push -f`.

**Comment by Alessandro-Battiato** [comment]
Thank you again for answering! 

There's definitely a lot of stuff I need to learn, but if my PR solved the issue then I'll just edit the description on this one and mention the issue number, but I surely need to apply soon enough what you suggested me to do with Git as I'm still having trouble understanding how everything works, even though I watched some videos and already worked on some projects but I can already see the difference between working by myself and contributing even for a small issue on a small project.

Regarding the formatting changes you're correct I did not change anything intentionally except for adding the noscript, I use prettier on my editor and I'm so used to how Prettier does everything automatically that I didn't even notice the changes!

Thanks for the precious advice, this first contribution brought me much value :)

File: index.xhtml [hunk 1]
```diff
@@ -8,13 +8,14 @@
...
   </head>
   <body>
+    <noscript>This website requires JavaScript.</noscript>
     <p>
       This is a live playground for rendering
       <a href="https://docutils.sourceforge.io/rst.html">reStructuredText</a>
-      into HTML entirely in the browser,
-      using <a href="https://pyodide.org">Pyodide</a>
-      to run <a href="https://docutils.sourceforge.io">docutils</a>.
-      Source code <a href="https://github.com/waldyrious/rst-playground">on GitHub</a>.
```
Comment (waldyrious) [inline index.xhtml:17]: Why did you re-wrap these lines? 🤔

File: script.js [hunk 2]
```diff
@@ -3,8 +3,18 @@ const outputFrame = document.getElementById("html-output");
 
 outputFrame.contentDocument.write("<!DOCTYPE html> Initializing...\n");
 
+// Checking if website uses WebAssembly
```
Comment (waldyrious) [inline script.js:6]:
```suggestion
// Check if the browser supports WebAssembly
```

File: script.js [hunk 3]
```diff
@@ -3,8 +3,18 @@ const outputFrame = document.getElementById("html-output");
...
+  if (!("WebAssembly" in window)) {
+    alert(
+      "This website requires WebAssembly because it depends on Pyodide to run docutils in the browser"
+    );
+  }
+}
+
 // Initialize Pyodide
 async function main() {
+  checkForWebAssembly();
```
Comment (waldyrious) [inline script.js:17]: I think this call should be at the very start of the `rstToHtml()` function, before the `let pyodide = await pyodideReadyPromise;` line. In fact, we should make the `try`/`catch` block start before the call to this function, so that this function could throw an exception if the browser doesn't have WebAssembly support, and then the error would be printed to the contents of the iframe using the code already implemented here: https://github.com/waldyrious/rst-playground/blob/f003e4fdd1fedca5982704576672c6fa6c01fb97/script.js#L49-L51

File: script.js [hunk 4]
```diff
@@ -40,13 +50,13 @@ async function rstToHtml() {
 
     // Override Docutils' default style, which adds a grey background to the body element.
     // We need to wait until the iframe's load event; see https://stackoverflow.com/a/13959836/266309.
-    outputFrame.addEventListener("load", event => {
+    outputFrame.addEventListener("load", (event) => {
```
Comment (waldyrious) [inline script.js:53]: I appreciate these syntax normalization changes, but please keep them in a separate commit (it can remain in this PR though)

**Comment by waldyrious** [comment]
Hey @Alessandro-Battiato! I've left some comments inline, but I didn't realize that @not-my-profile had already reviewed the PR. Sorry for the duplication! In any case, it can be helpful to use those inline comments as separate threads, to allow the various suggestions to be discussed individually without a lot of quoting back and forth :)

File: index.xhtml (hunk 1 above)
Comment (waldyrious) [reply index.xhtml:17]: Ah, I see you responded to this above:

> I did not change anything intentionally except for adding the noscript, I use prettier on my editor and I'm so used to how Prettier does everything automatically

It's probably a good idea to use something like Prettier, yeah. I'll open a new issue to track adding a tool like that to the repository, so that everybody can follow the same standard and future PRs won't have to deal with formatting changes mixed in :)

File: script.js (hunk 4 above)
Comment (waldyrious) [reply script.js:53]: Oh, I see these were done by Prettier as you described above. Let's ignore them for the time being, and once the PR is ready to merge we can split them into a separate commit. Don't worry about that for now.

File: script.js (hunk 2 above)
Comment (Alessandro-Battiato) [reply script.js:6]: Thanks I need to learn to use the proper terminologies as well :)

File: script.js (hunk 4 above)
Comment (Alessandro-Battiato) [reply script.js:53]: Indeed I apologize, as I said I didn't even notice these changes as I'm too used to Prettier fixing everything on its' own

File: script.js (hunk 3 above)
Comment (Alessandro-Battiato) [reply script.js:17]: Alright thanks for the tips! I'll fix everything as soon as I can :)

I apologize though for the confusion on my side but I have a question, you're saying that the call to the checkForWebAssembly function should be moved right before the let pyodide = await pyodideReadyPromise inside the rstToHtml function, but what do you mean when you say that we should make the try catch block start before the call of the check function?

Because If I understand correctly, once I move the call for the function before the pyodide variable, then the try catch block will never start first because I'm invoking the function right on top of the rstToHtml function's scope, so I'm a bit confused as to what you meant

Again thanks for all the tips and sorry if I was not clear enough trying to explain my doubt :)

File: index.xhtml (hunk 1 above)
Comment (waldyrious) [reply index.xhtml:17]: Created #26 to track this.

File: script.js (hunk 4 above)
Comment (waldyrious) [reply script.js:53]: No worries. As I said, we can deal with these formatting changes once the PR is ready to be merged. For now we can just leave them be :)

File: script.js (hunk 3 above)
Comment (waldyrious) [reply script.js:17]: No problem, I realize that I suggested multiple changes at once, and wasn't sufficiently explicit about that. I meant that in addition to moving the `checkForWebAssembly();` line to be before the `let pyodide = await pyodideReadyPromise;` line, the `try {` line should also be moved to be above the `checkForWebAssembly();` line. So the structure would be something like this:

```js
async function rstToHtml() {
  try {
    checkForWebAssembly();
    let pyodide = await pyodideReadyPromise;
    // ...
  } catch (err) {
    // ...
  }
}
```

Does this make sense?

Comment (Alessandro-Battiato) [reply script.js:17]: Yes it makes sense, I'm editing the code and will soon commit everything :)

File: script.js [hunk 5]
```diff
@@ -3,6 +3,15 @@ const outputFrame = document.getElementById("html-output");
 
 outputFrame.contentDocument.write("<!DOCTYPE html> Initializing...\n");
 
+// Check if the browser supports WebAssembly
+function checkForWebAssembly() {
+  if (!("WebAssembly" in window)) {
+    alert(
+      "This website requires WebAssembly because it depends on Pyodide to run docutils in the browser"
+    );
```
Comment (waldyrious) [inline script.js:11]: Ok, so now that this function is being called within a try/catch block, we can change this `alert()` to a [`throw()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw), as I suggested [earlier](https://github.com/waldyrious/rst-playground/pull/25#discussion_r1070421225).

That way, the error message will be cleanly printed into the output iframe, rather than opening an alert message, which I think makes the experience less jarring for the user.

Could you make this change, please?

File: script.js (hunk 2 above)
Comment (waldyrious) [reply script.js]: I'm marking this thread as resolved, as the change has been applied, to make the PR's discussion easier to follow.

File: script.js (hunk 3 above)
Comment (waldyrious) [reply script.js]: Great! I'll mark this thread as resolved so we can focus on the remaining ones.

File: script.js (hunk 5 above)
Comment (Alessandro-Battiato) [reply script.js:11]: My bad for not understanding this before!

I applied the changes but when I disable webassembly on my browser, the message I implemented in the function checkForWebAssembly is not displayed anywhere in the console, even after updating the alert code part as a throw, basically this is how I've updated the function: 

function checkForWebAssembly() {
  if (!("WebAssembly" in window)) {
    throw new Error(
      "This website requires WebAssembly because it depends on Pyodide to run docutils in the browser"
    );
  }
}

And everything else is still the same, though this is what you meant if I'm correct? Or am I missing something?

Comment (waldyrious) [reply script.js:11]: The message should be printed in the HTML page itself, in the output iframe (to the right of the input textbox), not in the console. That's because the `throw` is happening inside the `try`/`catch` block, and the catch block has code precisely to handle exceptions, rather than letting them propagate to the browser:

https://github.com/waldyrious/rst-playground/blob/f003e4fdd1fedca5982704576672c6fa6c01fb97/script.js#L48-L52

Doesn't the iframe show anything when you load the page with WebAssembly turned off?

Comment (waldyrious) [reply script.js:11]: Sure, you can commit and push your changes if it works for you locally. I can't guarantee that that version of the PR will be final, though — I want to try this locally as well, and I may suggest additional changes depending on the experience.

**Comment by waldyrious** [comment]
Ok, I've pushed a new commit (6ca22d5) with a proposal for an alternative to the `<noscript>` approach to tell the user that JavaScript is needed to run the site.

Instead of inserting a paragraph at the top of the page and leaving the controls (textarea, button) available for interaction but unable to produce any effect, I decided to disable those elements, and activate them via JavaScript. Furthermore, the user message is provided as a placeholder in the textarea, which is where they would go to if they wanted to use the conversion functionality.

These changes ensure that the layout is kept the same on both states, improve the probablility that the user will notice the "JavaScript needed" message (since it is right in the place where they'd normally try to interact with the site), and overall make the site work more intuitively. @Alessandro-Battiato and @not-my-profile, please check this behavior and let me know what you think!

The one downside I have noticed is that Firefox seems to be caching the disabled/enabled state, so disabling JS and performing a simple reload will have those elements enabled, but incapable of functioning. A hard refresh (Ctr+Shift+R) is required to actually load the page anew. This fortunately doesn't happen in Brave (a Chromium-based browser), and this situation should be rare in the first place, so I suppose it's not a big deal. But if anyone has ideas on how to bypass the browser's caching of element states, let me know!

**Comment by Alessandro-Battiato** [comment]
> Ok, I've pushed a new commit ([6ca22d5](https://github.com/waldyrious/rst-playground/commit/6ca22d5e77f33b7c47d19a360c125e7ecb4ed696)) with a proposal for an alternative to the `<noscript>` approach to tell the user that JavaScript is needed to run the site.
> 
> Instead of inserting a paragraph at the top of the page and leaving the controls (textarea, button) available for interaction but unable to produce any effect, I decided to disable those elements, and activate them via JavaScript. Furthermore, the user message is provided as a placeholder in the textarea, which is where they would go to if they wanted to use the conversion functionality.
> 
> These changes ensure that the layout is kept the same on both states, improve the probablility that the user will notice the "JavaScript needed" message (since it is right in the place where they'd normally try to interact with the site), and overall make the site work more intuitively. @Alessandro-Battiato and @not-my-profile, please check this behavior and let me know what you think!
> 
> The one downside I have noticed is that Firefox seems to be caching the disabled/enabled state, so disabling JS and performing a simple reload will have those elements enabled, but incapable of functioning. A hard refresh (Ctr+Shift+R) is required to actually load the page anew. This fortunately doesn't happen in Brave (a Chromium-based browser), and this situation should be rare in the first place, so I suppose it's not a big deal. But if anyone has ideas on how to bypass the browser's caching of element states, let me know!

I've tried the new functionalities on my machine, still using Firefox though so the elements are in fact enabled and not functioning, but still from a user perspective I am "forced" even before trying to type anything to read the placeholder which warns me that JavaScript is needed in order for the website to work properly, thus giving immediately the user the answer to the question " why won't it work".
But the placeholder option could probably have in my opinion a downside as compared to the noscript options, which happens in an extremely rare scenario where the user is plain ignoring every warning and every detail of the page and only writing inside the textarea without even reading the placeholder.
Considering though this is indeed a rare behaviour, I like this option as well, but regarding the firefox problem I'm sorry to say that I have no idea how to solve that :(

**Comment by waldyrious** [comment]
> But the placeholder option could probably have in my opinion a downside as compared to the noscript options, which happens in an extremely rare scenario where the user is plain ignoring every warning and every detail of the page and only writing inside the textarea without even reading the placeholder, making the latter disappear and having "no clue" as to what could be the cause of the problem that has broken the website.

That's a great point, I hadn't considered it! However, it shouldn't be an issue because besides adding the placeholder, the textarea is also made disabled by default, so the user should not be able to type into it. I just tested this in Brave, and even if the textarea has content, upon reloading the page with JS off, the textarea content is replaced with the placeholder (and the textarea is, as expected, disabled, so no new content can be added to make the placeholder go away).

IMO the fact that this issue can happen in Firefox is a bug _in Firefox_ that we should consider working around; but we shouldn't distort the base implementation just to accomodate one browser that behaves awkwardly. Let's wait for @not-my-profile's thoughts, and if none of us comes up with a workaround, I suggest we open an issue to track this problem and leave its resolution for a future PR.

**Comment by waldyrious** [comment]
Friendly ping to @not-my-profile :)

**Comment by not-my-profile** [comment]
I cannot reproduce the Firefox bug you describe with Firefox 112.0.2.

Anyway I don't think adding this message in an input placeholder is a good idea because of accessibility reasons ... screen readers may not read placeholder text. So I think it's better to go with the `<noscript>` tag as I initially suggested.

**Comment by waldyrious** [comment]
Fair enough. I would still suggest the `<noscript>` message to be placed within the opening paragraph (bolded, perhaps, to call attention to it), and the rest of the changes (disabled form elements, placeholder text) could still be present as additional hints. Let me push a new commit with these changes.

**Comment by waldyrious** [comment]
Update: I re-added the `<noscript>`, with the content wrapped in a `<mark>` element to make it stand out. I actually went back on the placeholder to declutter the interface and avoid duplication of the warning.

I still left the activation of the form and apparition of the instructional placeholder to happen via script, so that they don't provide mixed messages to the user who has JS disabled.

Here's what it looks like without JS:

image

...and with JS:

image

**Comment by waldyrious** [comment]
Hmm. I see your point, but IMO with the bold and the `<mark>` syntax, there's enough semantic and visual signaling of this notice's importance, so I'll keep the current approach that's less disruptive in layout terms.

**Comment by waldyrious** [comment]
Ok, I've rebased the branch to split the changes into three atomic commits, and force-pushed it to update the PR. One of them is the code formatting, slightly adjusted to remove the line re-wrapping in `index.xhtml` (since the current wrapping was chosen deliberately), and to add a new `'`→`"` change in `script.js`. These differences can be seen [here](https://github.com/waldyrious/rst-playground/compare/65d2c7dadf2827eaf0668d15e8fe503e12e35733..99cab355f30685151fc62f4fe15dbb13b7c5cf0e).

As for the JS and WebAssebly support checks, nothing changed in the combined diff compared to the previous version of the PR before the force-push, but the various commits were combined and re-split into two commits, one for each tech being checked. Your authorship of the commits was preserved, @Alessandro-Battiato :)

Merging now — thanks both for the contribution and review!

File: script.js (hunk 4 above)
Comment (waldyrious) [reply script.js:58]: The commit splitting was handled as described [below](https://github.com/waldyrious/rst-playground/pull/25#issuecomment-1537535729).
